provider/clc: CenturyLink Cloud Provider

ack opened the pull request:
initial PR for provider integration, ready for feedback.
- driver
- docs
- example config
- pending acc test account

phinze commented on builtin/providers/clc/resource_clc_group.go:
```diff
@@ -0,0 +1,118 @@
+		log.Printf("[INFO] Using EXISTING group: %v => %v", name, m[name])
+		d.SetId(m[name])
+		return nil
+	}
```
A bit more Go-style way to do this would be:

```go
if id, ok := m[name]; ok {
  log.Printf("...")
  d.SetId(id)
  return nil
}

phinze replied:
Ah and actually we shouldn't have this "use existing group" behavior in Create - it's inconsistent with the way resource creates generally work in Terraform. We'll implement import functionality as a first class feature. Attempting to create a resource that already exists should result in an error.

ack replied:
Part of the intent behind this is the relative inconvenience of working with server groups on clc. In particular, I was trying to allow the user to refer to a group by name rather than the clunkier guid. If instead the code errors on existing group in create, that would more or less require the user to either a) manage groups explicitly or b) book-keep guids. Not sure if the intention comes across through this description. But in the end, if it runs against convention, I'll remove it.

phinze replied:
Thanks for the extra context here.

If you want to enable users who are not managing server groups in Terraform to be able to reference them _from_ Terraform, it seems to me that the `name` vs `guid` question is more of an issue in the referencing resources rather than this group resource itself.

The behavior you get by rescuing this error is a "poor man's `terraform import`". If you assume at some future date that a user will be able to do something like `terraform import clc_group.mygroup "groupfoo"` - this "pull group under management on collision error during create" behavior ends up creating confusion.

Does that make sense?

ack replied:
yes. i'll have it error and await the import directive. thanks for clarifying.

ack replied:
hmm, so I revisited this problem and it's a bit thorny. here are the details:

- there's no way to create a new toplevel group and every group needs a parent
- parent_group_id is required on create
- groups don't have any uniqueness by attribute. eg. you can have multiple groups named foo (under the same parent). so it's not possible to create a resource (with some name) that already exists, a duplicate will always be created. 
- the tests may need a hard-coded guid
- import is not yet available

putting this altogether, it seems to me that having a poor man's import would be ideal. it's entirely possible I'm not understanding the intended pattern in the conventional way. would it be so horrible to break the rules here?

phinze replied:
Ah yes - thorny indeed!

I do see the merits in argument you're making, but the fact that groups don't have any uniqueness constraint makes me worry. What would the behavior be when there are two groups named "foo"?

Anyways, all this said - I'm willing to let this behavior stand for the initial implementation. We can let the implementation evolve with more usage.

ack replied:
Ya, there would be 2 groups with the same name and different IDs. Kinda
silly but that's how it works. I'll tidy up the rest of the group bits and
we can revisit later.
On Mar 2, 2016 3:58 PM, "Paul Hinze" <notifications@github.com> wrote:

> In builtin/providers/clc/resource_clc_group.go
> <https://github.com/hashicorp/terraform/pull/4893#discussion_r54814841>:
>
> > +}
> > +
> > +func resourceCLCGroupCreate(d *schema.ResourceData, meta interface{}) error {
> > +	client := meta.(*clc.Client)
> > +	dc := d.Get("location_id").(string)
> > +	m, err := dcGroups(dc, meta)
> > +	if err != nil {
> > +		return fmt.Errorf("Failed pulling groups in location %v - %v", dc, err)
> > +	}
> > +	name := d.Get("name").(string)
> > +	// use an existing group if we have one
> > +	if m[name] != "" {
> > +		log.Printf("[INFO] Using EXISTING group: %v => %v", name, m[name])
> > +		d.SetId(m[name])
> > +		return nil
> > +	}
>
> Ah yes - thorny indeed!
>
> I do see the merits in argument you're making, but the fact that groups
> don't have any uniqueness constraint makes me worry. What would the
> behavior be when there are two groups named "foo"?
>
> Anyways, all this said - I'm willing to let this behavior stand for the
> initial implementation. We can let the implementation evolve with more
> usage.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/terraform/pull/4893/files#r54814841>.
>

fatmcgav replied:
@ack

it might not be optimal, but I worked around this by using the top-level DC group… 

So `parent = "GB3 Hardware”`.

Could that be a sensible fall back?

Cheers
Gav

> On 2 Mar 2016, at 23:35, albert <notifications@github.com> wrote:
> 
> In builtin/providers/clc/resource_clc_group.go <https://github.com/hashicorp/terraform/pull/4893#discussion_r54812368>:
> 
> > +}
> > +
> > +func resourceCLCGroupCreate(d *schema.ResourceData, meta interface{}) error {
> > +	client := meta.(*clc.Client)
> > +	dc := d.Get("location_id").(string)
> > +	m, err := dcGroups(dc, meta)
> > +	if err != nil {
> > +		return fmt.Errorf("Failed pulling groups in location %v - %v", dc, err)
> > +	}
> > +	name := d.Get("name").(string)
> > +	// use an existing group if we have one
> > +	if m[name] != "" {
> > +		log.Printf("[INFO] Using EXISTING group: %v => %v", name, m[name])
> > +		d.SetId(m[name])
> > +		return nil
> > +	}
> hmm, so I revisited this problem and it's a bit thorny. here are the details:
> 
> there's no way to create a new toplevel group and every group needs a parent
> parent_group_id is required on create
> groups don't have any uniqueness by attribute. eg. you can have multiple groups named foo (under the same parent). so it's not possible to create a resource (with some name) that already exists, a duplicate will always be created.
> the tests may need a hard-coded guid
> import is not yet available
> putting this altogether, it seems to me that having a poor man's import would be ideal. it's entirely possible I'm not understanding the intended pattern in the conventional way. would it be so horrible to break the rules here?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/hashicorp/terraform/pull/4893/files#r54812368>.
>

phinze commented on builtin/providers/clc/resource_clc_group.go:
```diff
@@ -0,0 +1,118 @@
+	parent := m[p]
+	if parent == "" {
+		return fmt.Errorf("Failed resolving parent group %s - %s", p, m)
+	}
```
similar to above, the second `ok` argument to a map lookup is generally used in situations like this:

```go
parent, ok := m[p]
if !ok {
  return fmt.Errorf("...")
}
```

phinze commented on builtin/providers/clc/resource_clc_group.go:
```diff
@@ -0,0 +1,118 @@
+func resourceCLCGroupUpdate(d *schema.ResourceData, meta interface{}) error {
+	// unimplemented
+	return nil
+}
```
If Update is not implemented, you should omit this function and set all the attributes to ForceNew. That way, Terraform will handle any changes by replacing the resource.

ack replied:
Thanks, there actually is something to implement, but it does expand scope. Operations you'd like to perform on a collection of servers (power, archive, restore, etc) are supported and I'll get to this next.

phinze replied:
Yep that's fine - we often start with a ForceNew only initial implementation and circle back with enhancements for updateable attributes.

phinze commented on builtin/providers/clc/resource_clc_group.go:
```diff
@@ -0,0 +1,118 @@
+	d.Set("name", g.Name)
+	d.Set("description", g.Description)
+	// need to traverse links?
+	//d.Set("parent_group_id", g.ParentGroupID)
```
Any reason this is still commented out?

ack replied:
Ya, there's no method in the supporting sdk yet. Will need to patch the supporting library before being able to use this method.

phinze replied:
:ok_hand:

phinze commented on builtin/providers/clc/resource_clc_group_test.go:
```diff
@@ -0,0 +1,102 @@
+    location_id = "WA1"
+    name = "foobar"
+    parent = "Default Group"
+}`
```
Super tiny nit - I try to keep these config blocks formatted per [hclfmt](https://github.com/fatih/hclfmt) style - two space indent and aligned equals.

phinze commented on builtin/providers/clc/resource_clc_load_balancer.go:
```diff
@@ -0,0 +1,124 @@
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+			},
```
Comment says optional but field is required? :grinning:

phinze commented on builtin/providers/clc/resource_clc_load_balancer.go:
```diff
@@ -0,0 +1,124 @@
+		if err == nil {
+			break
+		}
+	}
```
Let's avoid the infinite loop here - we have a helper for this that makes it easier - check out [`resource.Retry`](https://github.com/hashicorp/terraform/blob/f09a5661cfec1239615a67d1143493dd7fa1b5f2/builtin/providers/aws/resource_aws_vpc.go#L311-L332).

phinze commented on builtin/providers/clc/resource_clc_load_balancer.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+func resourceCLCLoadBalancerUpdate(d *schema.ResourceData, meta interface{}) error {
+	//client := meta.(*clc.Client)
```
Stray comment line here

phinze commented on builtin/providers/clc/resource_clc_load_balancer.go:
```diff
@@ -0,0 +1,124 @@
+	id := d.Id()
+
+	if d.HasChange("name") {
+		d.SetPartial("name")
```
Since all the fields are updated in one API call, partial state handling doesn't matter - you can drop the `d.SetPartial` calls from this function.

phinze commented on builtin/providers/clc/resource_clc_load_balancer_pool.go:
```diff
@@ -0,0 +1,189 @@
+
+	s1 := d.Get("method").(string)
+	m := lb.LeastConn
+	if s1 == fmt.Sprintf("%v", lb.RoundRobin) {
```
Any reason for the `%v` here?

phinze commented on builtin/providers/clc/resource_clc_load_balancer_pool.go:
```diff
@@ -0,0 +1,189 @@
+		return err
+	}
+
+	// HACK: create does not allow pegging nodes so do an immediate update
```
No worries, this is a common pattern! :)

You might be able to just call through to your Update function in this file, a common resource pattern is:

```
func myResourceCreate(...) {
  ...
  // call update to catch any fields that can't be set during create
  return myResourceUpdate(...)
}
func myResourceUpdate(...) {
  ...
  // call read to refresh any fields that can't be read until after creation
  return myResourceRead(...)
}
```

phinze commented on builtin/providers/clc/resource_clc_load_balancer_pool.go:
```diff
@@ -0,0 +1,189 @@
+	pool, err := client.LB.GetPool(dc, lbid, id)
+	if err != nil {
+		return fmt.Errorf("Failed fetching pool under %v/%v: %v", dc, lbid, err)
+	}
```
Need to either recognize a NotFound / 404 error here and `d.SetId(""); return nil` or implement Exists() on the resource so you can assume the resource exists by the time you get to `Read()`.

Otherwise when a resource is deleted outside of Terraform we can't make progress because we keep getting the 404 error.

phinze commented:
Hello hello! I began a detailed code review - have to step AFK for now, but I shall return to complete it!

Feel free to follow up on any questions or concerns. :+1:

ack commented:
Thanks for following through and the detailed feedback! Will hop on these
revisions and comment inline.
On Feb 27, 2016 11:16 AM, "Paul Hinze" <notifications@github.com> wrote:

> Hello hello! I began a detailed code review - have to step AFK for now,
> but I shall return to complete it!
>
> Feel free to follow up on any questions or concerns. [image: :+1:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/terraform/pull/4893#issuecomment-189704476>.
>

ack commented:
Again, thanks for the extra set of eyes, style pointers and feedback. There are some pending bits I need to provide and I'd like your thoughts on a few minor points. Re: err-in-create-on-existing as some help with one of the tests.

ack commented:
This test has a step that is repeatedly failing. It is testing an update operation and after the update, re-reading my resource fails with a "Not found" (in the resource tree)

https://github.com/ack/terraform/blob/9c17a2dbc8255924878e25d1ba24a6319b8b185e/builtin/providers/clc/resource_clc_load_balancer_test.go#L38

The test errors (I think) here: 
https://github.com/ack/terraform/blob/9c17a2dbc8255924878e25d1ba24a6319b8b185e/helper/resource/testing.go#L360

Any clue what I'm doing wrong here or pointers to debugging it?

fatmcgav commented:
@ack Would be great to get the CLC provider into Terraform core... 

Could you take a look at https://github.com/CenturyLinkCloud/terraform-provider-clc/issues/4 and https://github.com/CenturyLinkCloud/terraform-provider-clc/issues/5 with the associated PR's before this gets merged though?

fatmcgav commented on examples/clc/variables.tf:
```diff
@@ -0,0 +1,19 @@
+  default = "achoi"
+}
+variable "clc_password" {
+  default = "T*5upfac"
```
@ack Are these "valid" credentials?

phinze commented:
@ack Trying to help out with that test question but I can't seem to get the base test to pass.

I think I've got the authentication sorted out, but I can't seem to figure out how to get around this 500 that the API is throwing:

```
=== RUN   TestAccLoadBalancerBasic
--- FAIL: TestAccLoadBalancerBasic (1.84s)
        testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

                * clc_load_balancer.acc_test_lb: Failed creating load balancer under WA1/acc_test_lb: http err: [500 Internal Server Error] - {"message":"An error has occurred."}
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/clc    1.852s
```

Any clues? My test account is `ZZ9F`

ack commented:
@phinze there were some set up steps to take on the account (create a network and a server) before the LB could be created. The complete message visible in our control portal was "Group could not be created: No private IP addresses." Also created an account for myself under ZZ9F to duplicate failed tests in the future. Apologies for the hiccup, please pull latest and give it a go now.

phinze commented:
Thanks @ack! Will give it another go.

> there were some set up steps to take on the account (create a network and a server) before the LB could be created.

Can we document whatever the required steps are to get from a fresh account to being able to run acceptance tests on the provider page?

phinze commented:
Looks like the load balancer "not found" message was just a typo: `s/clc_load_balancher.acc_test_lb/clc_load_balancer.acc_test_lb/`

Easy fix! :grinning:

ack commented:
added an update and some cleanup on groups. thx for the typo catch btw.

phinze commented:
Been working on getting this merge-able, and I believe I'm nearly there.

Here's my branch: https://github.com/hashicorp/terraform/tree/phinze/clc

In the latest commit on that branch I:

 * Update `resource.Retry` invocations for new API
 * Vendor the CLC SDK
 * Update the provider client initialization to match the API of the latest SDK

But now I can't seem to get the load balancer tests to pass:

```
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccGroupBasic
--- PASS: TestAccGroupBasic (80.35s)
=== RUN   TestAccLBPoolBasic
--- FAIL: TestAccLBPoolBasic (345.72s)
        testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

                * clc_load_balancer.acc_test_lbp: Failed creating load balancer under WA1/acc_test_lb: http err: [500 Internal Server Error] - {"message":"An error has occurred."}
=== RUN   TestAccLoadBalancerBasic
--- FAIL: TestAccLoadBalancerBasic (1.69s)
        testing.go:148: Step 0 error: Error applying: 1 error(s) occurred:

                * clc_load_balancer.acc_test_lb: Failed creating load balancer under WA1/acc_test_lb: http err: [500 Internal Server Error] - {"message":"An error has occurred."}
=== RUN   TestAccPublicIPBasic
kjkj


--- PASS: TestAccPublicIPBasic (824.65s)
=== RUN   TestAccServerBasic
--- PASS: TestAccServerBasic (480.40s)
FAIL
exit status 1
FAIL    github.com/hashicorp/terraform/builtin/providers/clc    1732.839s
```

@ack Any idea what might be going on here?

ack commented:
Ya, it's a known bug. The workaround was the mysterious ADMIN server that was in your account that I think you tore down. Basically, a server is required in the DC before a LB can be provisioned. I neglected to mention that the last time. If an operation failed w/ no helpful error message, you're likely to get a better one in the UI if you try the same operation.There's a system VPN server in IL1 that we can use permanently. 

I've pulled your changes, updated the tests to point to IL1, merged the latest sdk changes, and rebased the branch down to a few commits.

phinze commented:
Cool thanks for the extra context there - that makes sense. This looks good to go! :rocket: 

> Basically, a server is required in the DC before a LB can be provisioned.

Can we follow up with a modification to the LB tests so they just create their own servers?

phinze commented:
( Going to land this as-is though so it makes it in .14 :+1: - great work on this! :grinning: )

ack commented:
awesome! thanks so much for all your support getting this in.
